src/transformers/generation_utils.py

```diff
@@ -1356,9 +1376,20 @@ def generate(
             if num_beam_groups is not None and num_beam_groups > 1:
                 raise ValueError("`num_beam_groups` not supported yet for constrained generation.")
 
+            final_constraints = []
+            if constraints is not None:
+                final_constraints = constraints
+            if force_words_ids is not None:
+                for word_ids in force_words_ids:
+                    is_seq_tensor = isinstance(word_ids[0], torch.LongTensor) and len(word_ids[0].size()) > 0
+                    if isinstance(word_ids[0], list) or is_seq_tensor:
+                        constraint = DisjunctiveConstraint(word_ids)
+                    else:
+                        constraint = PhrasalConstraint(word_ids)
+                    final_constraints.append(constraint)
```

Comment:
It seems to me that we can keep everything like the old code and prevent complexitying further all those ifs by simply doing something like:

```python
forced_word_ids is not None:
    if constraints is None:
      constraints = []
    ....
    contraints.append(WhateverConstraint())
```
**before** all the ifs. Is there any reason why this wouldn't be sound ?

Reply:
Btw, this is the only place I think we should deal with many types and type checking. Everything afterwards should be a known type to prevent all kinds of ifs everywhere.

Reply:
I thought the new changes fulfilled the suggestion by dumping the type checks in this part of the code and not doing much type checks later down the line, but it seems I have misunderstood your comments to begin with. 

I'm not sure what you mean by "keeping everything like the old code" and "all the ifs". I don't quite understand the suggestion from:
```python
if forced_word_ids is not None:
    ...
if constraints is not None:
    ...
```
to 
```python
if forced_word_ids is not None:
    if constraints is None:
        ...
```

Reply:
I meant something like this:

```python
if forced_word_ids is not None:
    if constraints is None:
        constraints = []
    for word_ids in force_words_ids:
        is_seq_tensor = isinstance(word_ids[0], torch.LongTensor) and len(word_ids[0].size()) > 0
        if isinstance(word_ids[0], list) or is_seq_tensor:
            constraint = DisjunctiveConstraint(word_ids)
        else:
             constraint = PhrasalConstraint(word_ids)
         constraints.append(constraint)

# No change here
 is_constraint_gen_mode = constraints is not None
is_greedy_gen_mode = (num_beams == 1) and (num_beam_groups == 1) and do_sample is False and constraints is None
is_sample_gen_mode = (num_beams == 1) and (num_beam_groups == 1) and do_sample is True and constraints is None
is_beam_gen_mode = (num_beams > 1) and (num_beam_groups == 1) and do_sample is False and constraints is None

....

constrained_beam_scorer = ConstrainedBeamSearchScorer(
    constraints=constraints,
    ... 
)

```

I realized GH didn't make the suggestion on the whole block of code since it's separate changes for it, that's why you couldn't get what I was suggesting.

Basically the idea is to get `constraints` to be treated as a single thing everywhere it's possible.
Since `force_word_ids` is creating a new constraint, well, we sanitize and create the objects early on, and then everything is simpler later on.

FYI the same strategy is applied to `logits_processor` and `stopping_criteria`.

This `generate` function is definitely quite complex and managing its complexity is definitely not easy while we're trying to expand its capabilities (which is a good thing ! )

Reply:
Right, I understand now. 

Sort of why I went with this direction was because I wanted to ***visibly separate*** the constrained beam search part of the huge generate function code. 

It's actually necessary than a design choice for `logits_processor` to be defined before branching out to the different generation modes since *all* generation modes use them, but constraints can only be used in `constrained_beam_search` mode and can't integrate into any other modes. 

Considering that the type checking and processing on the `constraints` & `forced_words_ids `is complex and not short with 20+ lines, I thought it'd be actually cleaner to deal with it under
```python
if is_constrained_mode:
    ... # actually do the type checks & other processing, since it can only be used here
    # kind of a designated sanctuary for all constraints-related logic, separate from the rest of 
    # the complexities of generate()
```

Reply:
Good point ! Thanks for explaining that